lib/stripe/prices.rb

```diff
@@ -0,0 +1,188 @@
+      validates :recurring_usage_type,      inclusion: { in: %w{ metered licensed } }, allow_nil: true
+      validates :tiers_mode,                inclusion: { in: %w{ graduated volume } }, allow_nil: true
+
+      validate :name_or_product_id
+      validate :recurring_aggregate_usage_must_be_metered, if: ->(p) { p.recurring_aggregate_usage.present? }
+      validate :recurring_interval_count_maximum, if: ->(p) { p.recurring_interval_count.present? }
+      validate :valid_constant_name, unless: ->(p) { p.constant_name.nil? }
+
+      # validations for when using tiered billing
+      validate :tiers_must_be_array, if: ->(p) { p.tiers.present? }
+      validate :billing_scheme_must_be_tiered, if: ->(p) { p.tiers.present? }
+      validate :validate_tiers, if: ->(p) { p.billing_scheme == 'tiered' }
+
+      def initialize(*args)
+        super(*args)
+        @currency = 'usd'
+        @lookup_key = @id.to_s
+        @recurring = (recurring || {}).symbolize_keys
+      end
+
+      # We're overriding a handful of the Configuration methods so that
+      # we find and create by lookup_key instead of by ID.  The ID is assigned
+      # by stripe and out of our control
+      def put!
+        if exists?
+          puts "[EXISTS] - #{@stripe_class}:#{@id}:#{stripe_id}" unless Stripe::Engine.testing
+        else
+          object = @stripe_class.create({:lookup_key => @lookup_key}.merge compact_create_options)
+          puts "[CREATE] - #{@stripe_class}:#{object}" unless Stripe::Engine.testing
+        end
+      end
+
+      # You can't delete prices, but you can transfer the lookup key to a new price
+      def reset!
+        object = @stripe_class.create(reset_options)
+        puts "[RESET] - #{@stripe_class}:#{object}" unless Stripe::Engine.testing
+      end
+
+      def exists?
+        stripe_object.presence
+      rescue Stripe::InvalidRequestError
+        false
+      end
+
+      def stripe_object
+        @stripe_class.list({lookup_keys: [@lookup_key]}).data.first.presence || nil
+      rescue Stripe::InvalidRequestError
+        nil
+      end
+
+      def stripe_id
+        @stripe_id ||= stripe_object.try(:id)
+      end
+
+      def recurring_interval
+        recurring[:interval]
+      end
+
+      def recurring_aggregate_usage
+        recurring[:aggregate_usage]
+      end
+
+      def recurring_usage_type
+        recurring[:usage_type]
+      end
+
+      def recurring_interval_count
+        recurring[:interval_count]
+      end
+
+      private
+      def recurring_aggregate_usage_must_be_metered
+        errors.add(:recurring_aggregate_usage, 'recurring[:usage_type] must be metered') unless (recurring_usage_type == 'metered')
+      end
+
+      def recurring_interval_count_maximum
+        time_unit = recurring_interval.to_sym
+
+        return unless recurring_interval_count.respond_to?(time_unit)
+        too_long = recurring_interval_count.send(time_unit) > 1.year
```

Comment:
This is elegant! I'm pretty paranoid about using `.send` though. What do you think about whitelisting `time_unit` to `%w[day week month year]` before calling send?

I know it's already getting validated on `validates_inclusion_of  :recurring_interval,` but it isn't obvious that if someone updates that validation in the future that it will break this line.

On the other hand, whitelisting `time_unit` means validating this value twice, so I'm not sure what the right answer is.

Reply:
I'm just pleased to avoid doing date math on my own!  The latest version consolidates the allowed time units into a const for validation and re-use here.  I have zero concerns about validating it twice if it means keeping send safer.